### ISSUE TITLE
Fused staticmethod is not unwrapped in __get__

### ISSUE THREAD

**Issue by encukou** [cython/cython#3614]
A normal Cython function gets wrapped in Python's native `staticmethod`:
```python
cimport cython 
@staticmethod 
def statmeth(int arg): 
    return arg 

print(statmeth) 
```
```
<staticmethod object at 0x7fbad8fcadd0>
```

However, fused functions are instead `cyfunction`, with the `__Pyx_CYFUNCTION_STATICMETHOD` flag set:
```python
cimport cython 
@staticmethod 
def statmeth(cython.integral arg): 
    return arg 

print(statmeth) 
```
```
<cyfunction statmeth at 0x7fbaa44ef140>
```

This code:
```c
    if (func->self || func->func.flags & __Pyx_CYFUNCTION_STATICMETHOD) {
        // Do not allow rebinding and don't do anything for static methods
        Py_INCREF(self);
        return self;
    }
```
means that a fused staticmethod's `__get__` will return the staticmethod unchanged. This is different from Python's `staticmethod.__get__`, which instead returns the original function.
```python
cimport cython 

class C:
    @staticmethod 
    def normal(arg): 
        return arg 

    @staticmethod 
    def fused(cython.integral arg): 
        return arg 

obj = C()

for f in obj.normal, obj.fused: 
    print('normal:', f) 
    print('get once:', f.__get__(obj)) 
    print('get twice:', f.__get__(obj).__get__(obj)) 
    print('call:', f.__get__(obj).__get__(obj)()) 
    print() 
```
```python
normal: <staticmethod object at 0x7fbaa4398c10>
get once: <cyfunction normal at 0x7fbaa4529bb0>
get twice: <bound method normal of 42>
call: 42

normal: <cyfunction fused at 0x7fbad8ee5c80>
get once: <cyfunction fused at 0x7fbad8ee5c80>
get twice: <cyfunction fused at 0x7fbad8ee5c80>
Traceback (most recent call last)
   ...
TypeError: Expected at least 1 argument, got 0
```

**Comment by scoder** [comment]
Did you try this with the latest master?

**Comment by scoder** [comment]
Could be one of these: #3381 #3106

**Comment by encukou** [comment]
Yes, I get this on current master. (Though the reason I'm digging is trying to get Cython 0.29 compatible with Python 3.9.)
Thanks for the pointers. I believe it's something else, though I admit I'm somewhat lost in the codebase.
It looks to me like `CYFUNCTION_STATICMETHOD` is not really used in regular functions. I'm not sure how it's being set for the fused ones.

**Comment by da-woods** [comment]
> However, fused functions are instead cyfunction, with the __Pyx_CYFUNCTION_STATICMETHOD flag set

That at least has to be true, since the function still needs to act as a fused function (e.g. it needs to be indexable tos select the specialization).

Possibly what should happen is for `__get__` to return a copy of the fused function without `__Pyx_CYFUNCTION_STATICMETHOD` set?

**Comment by scoder** [comment]
Yeah, that seems worth a try. It's a bit costly, though. I think, for the specific case of static methods, we could just cache the instance and always return it from that point on.

**Comment by encukou** [comment]
> That at least has to be true, since the function still needs to act as a fused function (e.g. it needs to be indexable tos select the specialization).

That means using `staticmethod` as a decorator works differently than calling it, as in:

```
cimport cython 
 
@classmethod 
def with_decorator(cython.numeric arg): 
    return arg 
 
def _func(cython.numeric arg): 
    return arg 
with_call = staticmethod(_func) 

with_decorator[int]
# with_call[int]  # error
```
Is that expected?

**Comment by da-woods** [comment]
I think it's probably unavoidable. It isn't ideal behaviour but having fused functions work as fused functions is more important than having them be perfectly consistent with Python functions.

Obviously a suitable fix would be an improvement, but it's hard to see what that fix would be.

**Comment by scoder** [comment]
> Is that expected?

Well, I don't see a way around this. `staticmethod` is implemented on Python side. I mean, ok, we could special case it inside of Cython compiled code, but then it would still behave differently when you run your code uncompiled (actually, I guess it already does that…) or call the decorator elsewhere or by other (not so obvious) means.

**Comment by encukou** [comment]
I guess the hyper-correct thing would be a `cython.staticmethod` decorator. But I'll let you weigh the pros and cons; I'm just making sure you're aware of the edge case.

**Comment by scoder** [comment]
> the hyper-correct thing would be a cython.staticmethod decorator

… and then users would forget to import that and would still end up with the Python one, but only in 43% of the cases. ¯\_(ツ)_/¯

**Comment by da-woods** [comment]
Looking at it again, I'm not sure there's a good reason for static fused methods to be special most of the time. Except in cdef classes they behave fine when just wrapped with `staticmethod`.

`classmethod` looks like it does have to stay special, so will probably have similar slight incompatibilities.

`__get__` does become quite expensive for `staticmethod` though since it has to copy a fairly complex structure instead of just incrementing a reference. It might be worth fixing it though - although I should probably understand it a little better first.
